File: zig/binary_tree_traverse.zig
const std = @import("std");
const print = std.debug.print;

const Node = struct {
    data: i32,
    left: ?*Node,
    right: ?*Node,
};

fn buildTree(allocator: std.mem.Allocator, depth: i32, counter: *i32) !?*Node {
    if (depth <= 0) return null;
    
    const node = try allocator.create(Node);
    node.data = counter.*;
    counter.* += 1;
    node.left = try buildTree(allocator, depth - 1, counter);
    node.right = try buildTree(allocator, depth - 1, counter);
    return node;
}

fn traverseInOrder(node: ?*Node) i32 {
    if (node == null) return 0;
    
    var sum: i32 = 0;
    sum += traverseInOrder(node.?.left);
    sum += node.?.data;
    sum += traverseInOrder(node.?.right);
    return sum;
}

fn freeTree(allocator: std.mem.Allocator, node: ?*Node) void {
    if (node == null) return;
    
    freeTree(allocator, node.?.left);
    freeTree(allocator, node.?.right);
    allocator.destroy(node.?);
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();
    
    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);
    
    if (args.len < 2) {
        print("Usage: binary_tree_traverse <depth>\n", .{});
        return;
    }
    
    const depth = try std.fmt.parseInt(i32, args[1], 10);
    var counter: i32 = 0;
    
    const start = std.time.nanoTimestamp();
    const root = try buildTree(allocator, depth, &counter);
    const sum = traverseInOrder(root);
    freeTree(allocator, root);
    const end = std.time.nanoTimestamp();
    
    const time_ms = @divFloor(end - start, 1_000_000);
    
    print("Zig: binary_tree_traverse({d}) = {d}\n", .{ depth, sum });
    print("Time: {d}ms\n", .{time_ms});
}

File: c/binary_tree_traverse.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

typedef struct Node {
    int data;
    struct Node* left;
    struct Node* right;
} Node;

Node* createNode(int data) {
    Node* node = malloc(sizeof(Node));
    node->data = data;
    node->left = NULL;
    node->right = NULL;
    return node;
}

Node* buildTree(int depth, int* counter) {
    if (depth <= 0) return NULL;
    
    Node* node = createNode((*counter)++);
    node->left = buildTree(depth - 1, counter);
    node->right = buildTree(depth - 1, counter);
    return node;
}

int traverseInOrder(Node* node) {
    if (node == NULL) return 0;
    
    int sum = 0;
    sum += traverseInOrder(node->left);
    sum += node->data;
    sum += traverseInOrder(node->right);
    return sum;
}

void freeTree(Node* node) {
    if (node == NULL) return;
    
    freeTree(node->left);
    freeTree(node->right);
    free(node);
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        printf("Usage: binary_tree_traverse <depth>\n");
        return 1;
    }
    
    int depth = atoi(argv[1]);
    int counter = 0;
    
    clock_t start = clock();
    Node* root = buildTree(depth, &counter);
    int sum = traverseInOrder(root);
    freeTree(root);
    clock_t end = clock();
    
    double time_ms = ((double)(end - start) / CLOCKS_PER_SEC) * 1000;
    
    printf("C: binary_tree_traverse(%d) = %d\n", depth, sum);
    printf("Time: %.0fms\n", time_ms);
    
    return 0;
}